### ISSUE TITLE
尝试复现知识蒸馏方法时教师网络输出与学生网络输出之间的loss训练时保持不变

### ISSUE THREAD

**Issue by reveLATONG** [TomTomTommi/LiteAnyStereo#6]
我尝试在lightstereo上复现论文中提到的知识蒸馏方法，采用高斯模糊、随机擦除、色彩抖动、过曝等方法对学生网络的输入进行了扰动，蒸馏代码如下
```
class distill_s2(nn.Module):
    def __init__(self, args):
        super(distill_s2, self).__init__()

        self.args = args
        self.teacher_model = Lightstereo(args)
        self.student_model = Lightstereo(args)

        self.perturber = StrongPerturbation(normalize_cfg=NORM_CONFIG, jitter_cfg=JITTER_CONFIG)

        for param in self.teacher_model.parameters():
            param.requires_grad = False

    def forward(self, data):
        image1, image2 = data['left'], data['right']

        perturbation_image1, perturbation_image2 = self.perturber(image1, image1)

        self.teacher_model.eval()

        with torch.no_grad():
            teacher_output = self.teacher_model({'left': image1, 'right': image2})

        student_output = self.student_model({'left': perturbation_image1, 'right': perturbation_image2})

        result = {'teacher_output': teacher_output, 'student_output': student_output}

        return result

    def get_loss(self, result, data):
        teacher_output = result['teacher_output']
        student_output = result['student_output']
        disp_loss, _ = self.student_model.get_loss(student_output, data)

        # teacher_norm = l2_normalize(teacher_output, dim=0)
        # student_norm = l2_normalize(student_output, dim=0)
        teacher_norm = teacher_output / (teacher_output.norm(p=2, dim=(2, 3), keepdim=True) + 1e-8)
        student_norm = student_output / (student_output.norm(p=2, dim=(2, 3), keepdim=True) + 1e-8)
        b, c, h, w = teacher_output.shape
        # cos_loss = 1 - torch.sum(teacher_norm * student_norm)/(h * w)
        cos_loss = 0.0
        for tn, sn in zip(teacher_norm, student_norm):
            cos_loss += 1 - torch.sum(tn * sn)/(h * w)

        loss = disp_loss + cos_loss
        print(f"disp loss:{disp_loss.item():#.6g} and cos loss:{cos_loss.item():#.6g}")
        # print("cos_loss:", cos_loss)

        loss_info = {'scalar/train/loss_disp': loss.item()}

        return loss, loss_info
```
训练过程中打印学生网络的loss以及蒸馏的cosine loss，发现师生网络的差异一直保持不变，是知识蒸馏的设计有问题吗？我应当只对输出的disp进行蒸馏还是需要对一些中间层进行蒸馏呢？

**Comment by zaowanzazhi** [comment]
请问数据扰动的代码可以参考一下吗，我看论文说在补充材料里面，但我没有找到

**Comment by TomTomTommi** [comment]
Hi, please see the settings of data augmentation:
```
class FlowAugmentor:
    def __init__(self, crop_size, min_scale=-0.2, max_scale=0.5, do_flip=True, yjitter=False,
                 saturation_range=[0.6, 1.4], gamma=[1, 1, 1, 1]):

        # spatial augmentation params
        self.crop_size = crop_size
        self.min_scale = min_scale
        self.max_scale = max_scale
        self.spatial_aug_prob = 1.0
        self.stretch_prob = 0.8
        self.max_stretch = 0.2

        # flip augmentation params
        self.yjitter = yjitter
        self.do_flip = do_flip
        self.h_flip_prob = 0.5
        self.v_flip_prob = 0.1

        # photometric augmentation params
        self.photo_aug = Compose([
            ColorJitter(
                brightness=0.8,  # ±80%
                contrast=0.8,  # ±80%
                saturation=[0.2, 1.8],  # Wide saturation range
                hue=0.3  # ≈ ±0.3 radians hue shift
            ),
            RandomGaussianBlur(kernel_size=5, sigma=(0.1, 2.0), p=0.5),
            AdjustGamma(0.7, 1.5)  # Strong gamma variation
        ])

        self.asymmetric_color_aug_prob = 0.2
        self.eraser_aug_prob = 0.1
```

**Comment by TomTomTommi** [comment]
```
class AdjustGamma:
    def __init__(self, gamma_min, gamma_max, gain_min=1.0, gain_max=1.0, p=1.0):
        self.gamma_min = gamma_min
        self.gamma_max = gamma_max
        self.gain_min = gain_min
        self.gain_max = gain_max
        self.p = p

    def __call__(self, sample):
        if random.random() > self.p:
            return sample
        gain = random.uniform(self.gain_min, self.gain_max)
        gamma = random.uniform(self.gamma_min, self.gamma_max)
        return functional.adjust_gamma(sample, gamma, gain)


class RandomGaussianBlur:
    def __init__(self, kernel_size=5, sigma=(0.1, 2.0), p=0.5):
        self.blur = GaussianBlur(kernel_size, sigma)
        self.p = p

    def __call__(self, img):
        if random.random() < self.p:
            return self.blur(img)
        return img
```

**Comment by zaowanzazhi** [comment]
非常感谢你的分享，你的无私感染着我们

**Comment by zaowanzazhi** [comment]
def compute_feature_loss(features_student_list, features_teacher_list, ignore_top_percent=0.10, scale_weights=None)，请问特征的输入是来自与cv吗，cv = self.cost_agg_2d(cv_3d, features_left)就是这个，谢谢

**Comment by TomTomTommi** [comment]
we use the features from fnet:

```
    def forward(self, left, right, max_disp=192, test_mode=False, kd_mode=False):
......

        features_left = self.fnet(left)
        features_right = self.fnet(right)

......

        if test_mode:
            return disp_up
        elif kd_mode:
            disp_linear = F.interpolate(disp, left.shape[2:], mode='bilinear', align_corners=False)
            return [disp_up, disp_linear * 4.], features_left, features_right
        else:
            disp_linear = F.interpolate(disp, left.shape[2:], mode='bilinear', align_corners=False)
            return [disp_up, disp_linear * 4.]
```

**Comment by zaowanzazhi** [comment]
好的，感谢

**Comment by ZhangHefu** [comment]
> Hi, thanks for your interest. Teacher model is frozen in stage 2. The code we use is as follows:
> 
> ```
> def compute_feature_loss(features_student_list, features_teacher_list, ignore_top_percent=0.10, scale_weights=None):
>     """
>     Compute feature distillation loss over multiple feature levels.
> 
>     Args:
>         features_student_list: list of tensors, each [B, C, H, W]
>         features_teacher_list: list of tensors, each [B, C, H, W]
>         ignore_top_percent: fraction of highest-loss pixels to ignore
>         scale_weights: optional list of per-scale weights
> 
>     Returns:
>         scalar loss
>     """
>     assert len(features_student_list) == len(features_teacher_list), "Feature lists must match in length"
> 
>     total_loss = 0.0
>     num_scales = len(features_student_list)
> 
>     if scale_weights is None:
>         scale_weights = [1.0] * num_scales
> 
>     for f_s, f_t, weight in zip(features_student_list, features_teacher_list, scale_weights):
>         # Normalize to compute cosine similarity
>         f_s = F.normalize(f_s, p=2, dim=1)  # [B, C, H, W]
>         f_t = F.normalize(f_t, p=2, dim=1)  # [B, C, H, W]
> 
>         # Cosine similarity map
>         cos_sim = (f_s * f_t).sum(dim=1)  # [B, H, W]
>         loss_map = 1 - cos_sim  # [B, H, W]
> 
>         # Flatten and compute top-k mask
>         B, H, W = loss_map.shape
>         loss_map_flat = loss_map.view(B, -1)
> 
>         top_k = int(ignore_top_percent * H * W)
>         sorted_loss, _ = torch.sort(loss_map_flat, dim=1, descending=True)
>         threshold = sorted_loss[:, top_k].unsqueeze(1)  # [B, 1]
>         mask = (loss_map_flat < threshold).float()  # [B, HW]
> 
>         # Masked loss per batch element
>         masked_loss = (loss_map_flat * mask).sum(dim=1) / mask.sum(dim=1).clamp(min=1.0)  # [B]
>         total_loss += weight * masked_loss.mean()
> 
>     return total_loss
> ```

How did you choose which features to include in 'features_student_list' and 'features_teacher_list'? I also noticed the hyper-para 'scale_weights'  — do you have any findings on how to set these weights for different feature levels that you could share?

**Comment by TomTomTommi** [comment]
Hi,

Thanks for your interest. For `features_student_list` and `features_teacher_list`, we simply use the feature maps at three resolutions: 1/16, 1/8, and 1/4.

Regarding `scale_weights`, we did not perform a dedicated ablation on this hyper-parameter. In our experiments, we simply left it as `None`

We also found that Stage 2 is not where most of the final performance gain comes from. From our perspective, Stage 3 plays a more important role in improving the final results.
